Step2 미션 제출합니다

iimpala opened the pull request:
피드백 주신 부분 반영해서 다시 PR 보냅니다:)

한가지 질문이 있는데 LottoResult를 변환하는 과정에서 해당 로직이 View레이어로 가야 할 것 같다고 피드백주셨는데
해당 로직이 View레이어로 넘어가면 view에서 domain의 Prize를 알게 되어 MVC패턴에 위배된다고 생각해서 컨트롤러에서 원시타입인 String으로 변환하려 한 것입니다. 그런데 지금은 단순 데이터만 변환한게 아니라 출력형식까지 컨트롤러에서 만들어주기 때문에 이런 피드백을 주신거라고 이해했습니다.

그렇다면 지금 controller에서 view로 넘어가야 하는 데이터가

- 각 등수별 당첨 개수 `result.getOrDefault(prize, 0L)`
- 각 등수별 매칭되는 숫자의 개수 `prize.getMatchCount()`
- 각 등수별 상금 `prize.getReward()`
- 등수를 구분할 수 있는 이름 `prize.name()`
- 전체 수익률 `lottoResult.getProfitRate(budget)`

정도로 생각하는데 사실상 Prize객체의 모든 정보가 넘어가야 한다고 생각이 됩니다.

이럴 경우에 원시타입으로 Prize를 포장한 PrizeDto와 같은 객체를 만들어서 넘겨주어야 할까요? 아니면 다른 좋은 방안이 있을까요?

- 이전 PR Link: https://github.com/next-step/java-lotto-kakao/pull/59

humbledude commented on src/main/java/lotto/domain/WinningLottoNos.java:
```diff
@@ -14,8 +14,8 @@ public WinningLottoNos( List<Integer> numbers, int bonusNumber) {
-        if ( this.lottoNumbers.isContains(lottoNumber) ) {
-            throw new IllegalArgumentException();
+        if ( lottoNumbers.isContains(lottoNumber) ) {
+            throw new IllegalArgumentException("잘못된 입력 입니다");
```
뭐가 잘못되었는지 구체적으로 알려주는게 좋겠어요~
현업에서도 수많은 exception 을 뿌릴텐데 그때마다 코드 보는거보다 로그가 친절하게 알려주는게 더 낫습니다

humbledude commented on src/main/java/lotto/input/ValidInputGuaranteer.java:
```diff
@@ -6,12 +6,13 @@
             return transformToObject(valueFromUser);
         } catch (Exception e){
-            System.out.println(e.getMessage());
+            System.out.println("적절하지 않은 입력 입니다.");
```
각 구현체에서 input validation 이 실패하면 Exception 을 던지고, 그 exception 에 포함된 메시지가 출력되게 하는게 좋은거 같습니다.

Yookibaek replied:
지적 감사합니다!

humbledude reviewed (approved):
수고 많으셨습니다~

1단계에서 거의 구현해 놓아서 2단계는 수월하게 진행된거 같아요~
자잘한 피드백 남겨두었으니 참고하시고,
다음 미션도 즐겁게 하시길 바랍니다~

humbledude commented on src/test/java/lotto/input/SelfLottoCountValidInputGuaranteerTest.java:
```diff
@@ -0,0 +1,12 @@
+    public void validInputTest(){
+        assertThat(new SelfLottoCountValidInputGuaranteer().transformToObject("3")).isEqualTo(3);
+    }
+}
```
https://blog.coderifleman.com/2015/04/04/text-files-end-with-a-newline/

파일끝 newline 이 필요한 이유에 대해서 나와있어요~
참고 하시고, 거기 웹스톰 설정이라고 된 부분이 인텔리j 설정이랑 같아요~
오늘 설정하시면 맥북 반납시까지 안전~

iimpala committed:
feat(LottoPurchaseBudget) 로또 비용차감 로직 작성

- 구입하려는 로또 개수를 받아 예산 안에 드는지 검증하는 로직 작성
- 로또 구입시 금액만큼 예산을 차감하는 로직 작성

iimpala committed:
feat(LottoTicketSeller) 수동 로또 구매 로직 작성

- 수동으로 로또 번호를 받아 구입하는 로직 추가

iimpala committed:
refactor(LottoPurchaseBudget): 에러 메시지 수정

공백 추가

iimpala committed:
refactor(Prize): Prize에서 view 의존성 제거

Prize클래스 안에 view에 대한 의존성이 있던 부분을 컨트롤러로 이동
- 로또의 결과를 String으로 변환하는 부분을 Prize에서 컨트롤러로 이동
- 로또 결과를 List<String>로 매핑하여 view에서 단순 출력만 할 수 있도록 수정

iimpala committed:
refactor(LottoNumber): CACHE 초기화 로직 수정

IntStream을 사용하여 캐시를 초기화 할 수 있도록 수정

iimpala committed:
refactor(LottoTicketSeller): 상수 추출

한 티켓에 들어가는 로또번호의 개수를 의미하는 숫자를 상수로 추출

iimpala committed:
refactor(LottoTicket): LottoTicket 리팩토링

- 상수와 인스턴스 변수 사이에 공백라인 추가
- 메소드명 변경 (compare -> countMatchedNumbers)

iimpala committed:
refactor(LottoResult): LottoResult 리팩토링

LottoResult가 일급 컬렉션의 성격을 띄도록 수정
- 결과 집계로직을 WinningLotto로 이전

iimpala committed:
refactor(LottoTicketSeller): LottoTicketSeller 정적 클래스로 전환

상태를 가지지 않는 객체의 생성을 하지 않는 방향으로 개선

iimpala committed:
refactor(WinningLotto): aggregateResult가 LottoResult를 직접 반환하도록 변경

iimpala committed:
refactor(LottoController): TicketDto 매핑 로직 수정

TicketDto를 매핑하는 메서드와 List<TicketDto>를 매핑하는 메서드를 분리

iimpala committed:
refactor(LottoPurchaseBudget): 로또 구입 개수 검증 로직 수정

개수가 아닌 가격을 비교하도록 변경

iimpala committed:
refactor(LottoResult): 수익률 계산 로직 수정

keySet() -> entrySet()

iimpala committed:
refactor(LottoTicketSeller): 메소드명 변경

일관성있는 네이밍을 위해 purchaseAutoLotto() -> purchaseAutoLottoTickets()로 변경

iimpala committed:
refactor(LottoTickets): LottoTickets 일급 컬렉션 생성

List<LottoTicket>을 감싼 LottoTickets 일급 컬렉션 작성

Rok93 reviewed (changes requested):
동현님 피드백 잘 반영해주셨습니다! 👍💯
질문 주신 내용에 제 생각 남겨두었고  추가적으로 코멘트 몇개 남겨두었으니 
확인해서 반영해주세요. 🙏
아마 다음에는 머지할 수 있을 것 같아요! 💪💪

iimpala committed:
refactor(LottoTickets): addAll 반환타입 변경

LottoTickets의 addAll 메소드 호출시 새로운 LottoTickets 객체를 반환하도록 변경

iimpala committed:
refactor(PrizeResultDto): 로또 당첨결과 출력 로직을 view 레이어로 위임

PrizeResultDto를 만들어 controller -> view로 당첨정보를 전달하도록 변경

Rok93 reviewed (approved):
동현님 피드백 반영 잘해주셨습니다. 👍💯
로또 미션은 충분히 잘 구현해주신 것 같아서, 이만 머지하도록하겠습니다. 👏👏
질문 주신 내용에는 제 생각 남겨두었으니 확인 부탁드릴게요. 🙂

그동안 고생많으셨어요. 🙇‍♂️
남은 미션도 화이팅입니다. 💪💪
